File: gaussian_basis/gaussian_basis/closed_shell_system.py
from typing import List, Tuple, Union
import numpy as np
from scipy.linalg import eigh
from . import extension
from .matrices import get_overlap_matrix, get_kinetic_matrix
from .matrices import get_nuclear_potential_matrix
from .matrices import get_two_electron_integrals_tensor
from .orbitals import *

class ClosedShellSystem:

    _orbitals_data: np.ndarray
    orbitals_count: int
    basis_func_count: int
    orbitals: np.ndarray
    overlap: np.ndarray
    kinetic: np.ndarray
    nuclear: np.ndarray
    hamiltonian: np.ndarray
    two_electron_integrals: np.ndarray
    energies: np.ndarray
    nuclear_configuration: List[List[Union[np.ndarray, int]]]

    def __init__(self,
                 number_of_electrons,
                 orbitals: Orbitals,
                 nuclear_configuration: List[List[Union[np.ndarray, int]]]
                 ):
        self.nuclear_configuration = nuclear_configuration
        self._orbitals_data = encode_orbitals_dict_as_array(orbitals.atomics)
        basis_func_count = self._orbitals_data.view(dtype=np.int64)[0]
        print('basis function count', basis_func_count)
        j = 0
        self.orbitals_count = number_of_electrons # orbitals.number_of_orbitals()
        self.orbitals = np.zeros(
            [self.orbitals_count, basis_func_count],
            dtype=np.double)
        for i, orbital in enumerate(orbitals.atomics):
            orbitals_dict = orbital.get_dictionary_of_parameters()
            for k, orbital_name in enumerate(orbitals_dict):
                multiplicity = min(
                    3, get_multiplicity_from_orbital_name(orbital_name))
                for m in range(multiplicity):
                    for basis_func in orbitals_dict[orbital_name]:
                        self.orbitals[
                            min(i + k + m, number_of_electrons-1), j]\
                            = basis_func['coefficient']
                        j += 1
        nuclear_locations = np.zeros(
            [3*len(nuclear_configuration)], dtype=np.double, order='C')
        charges = np.zeros(
            [len(nuclear_configuration)], dtype=np.intc, order='C')
        for i, e in enumerate(nuclear_configuration):
            nuclear_locations[3*i: 3*i + 3] = np.copy(e[0])
            charges[i] = e[1]
        print("Nuclear locations: ", nuclear_locations)
        print("charges: ", charges)
        print("Constructing matrices...")
        self.overlap = np.zeros(
            [basis_func_count, basis_func_count], dtype=np.double, order='C')
        self.kinetic = np.zeros(
            [basis_func_count, basis_func_count], dtype=np.double, order='C')
        self.nuclear = np.zeros(
            [basis_func_count, basis_func_count], dtype=np.double, order='C')
        self.two_electron_integrals = np.zeros(
            4*[basis_func_count], dtype=np.double, order='C'
        )
        self.basis_func_count = basis_func_count
        print("Constructing overlap matrix...")
        extension.compute_overlap(self.overlap, self._orbitals_data)
        print("Constructing kinetic matrix...")
        extension.compute_kinetic(self.kinetic, self._orbitals_data)
        print("Constructing nuclear matrix...")
        extension.compute_nuclear(
            self.nuclear, nuclear_locations, charges, self._orbitals_data)
        print("Constructing two electron integrals matrix...")
        extension.compute_two_electron_integrals(
            self.two_electron_integrals, self._orbitals_data
        )
        # import matplotlib.pyplot as plt
        # plt.imshow(self.overlap)
        # plt.title("Overlap matrix")
        # plt.show(); plt.close()
        # plt.title("Kinetic matrix")
        # plt.imshow(self.kinetic)
        # plt.show(); plt.close()
        # plt.title("Nuclear matrix")
        # plt.imshow(self.nuclear)
        # plt.show(); plt.close()
        self.hamiltonian = self.kinetic + self.nuclear
        self.energies = np.zeros([self.orbitals_count])


    def solve(self, iter_count):
        orbitals = self.orbitals
        energies = self.energies
        two_electron_integrals = self.two_electron_integrals
        h = self.hamiltonian
        # orbital_norm_factor = 1
        for i in range(iter_count):
            repulsion = 2.0*np.einsum('ijkl,mk,ml->ij',
                                      two_electron_integrals,
                                      *2*[orbitals],
                                      optimize='greedy',
                                      )
            exchange = np.einsum('ijkl,mj,ml->ik',
                                 two_electron_integrals,
                                 *2*[orbitals],
                                 optimize='greedy',
                                 )
            fock = h + repulsion - exchange
            energies, eigenvectors = eigh(
                fock, b=self.overlap,
                subset_by_index=[0, self.orbitals_count-1])
            orbitals = eigenvectors.T
            # orbital_norm_factor = \
            #     np.sqrt(np.sum(orbitals @ (self.overlap @ orbitals.T)))
            # orbitals = orbitals / orbital_norm_factor
        self.orbitals = orbitals
        self.energies = energies
        return energies, orbitals

    def get_kinetic_energy(self):
        return 2.0*np.einsum('ij,ni,nj->', self.kinetic,
                             * 2*[self.orbitals])

    def get_nuclear_potential_energy(self):
        return 2.0*np.einsum('ij,ni,nj->', self.nuclear,
                             * 2*[self.orbitals])

    def get_repulsion_exchange_energy(self):
        repulsion_e = np.einsum('ijkl,ni,nj,mk,ml->',
                                self.two_electron_integrals,
                                * 4*[self.orbitals],
                                optimize='greedy')
        exchange_e = np.einsum('ijkl,ni,mj,nk,ml->',
                               self.two_electron_integrals,
                               * 4*[self.orbitals],
                               optimize='greedy')
        return 2.0*repulsion_e - exchange_e

    def get_nuclear_configuration_energy(self):
        val = 0.0
        if len(self.nuclear_configuration) > 1:
            for i, c_i in enumerate(self.nuclear_configuration):
                for j in range(i+1, len(self.nuclear_configuration)):
                    c_j = self.nuclear_configuration[j]
                    r_i, q_i = c_i
                    r_j, q_j = c_j
                    r = np.linalg.norm(r_i - r_j)
                    val += q_i*q_j/r
            return val
        else:
            return 0.0

    def get_total_energy(self):
        return (self.get_kinetic_energy()
                + self.get_nuclear_potential_energy()
                + self.get_repulsion_exchange_energy()
                + self.get_nuclear_configuration_energy())


class ClosedShellSystemFromPrimitives:

    orbitals: np.ndarray
    orbitals_count: np.ndarray
    overlap: np.ndarray
    kinetic: np.ndarray
    nuclear: np.ndarray
    h: np.ndarray
    two_electron_integrals: np.ndarray
    energies: np.ndarray
    nuclear_configuration: List[List[Tuple[np.ndarray, int]]]

    def __init__(self, **kw):
        primitives = kw['primitives']
        self.orbitals = kw['orbitals']
        nuclear_config = kw['nuclear_config']
        self.nuclear_configuration = nuclear_config
        self.orbitals_count = self.orbitals.shape[0]
        self.energies = np.zeros([self.orbitals_count])
        if 'use_ext' in kw and kw['use_ext']:
            self.init_using_ext(primitives)
        else:
            self.overlap = get_overlap_matrix(primitives)
            self.kinetic = get_kinetic_matrix(primitives)
            self.nuclear = get_nuclear_potential_matrix(primitives,
                                                        nuclear_config)
            self.h = self.kinetic + self.nuclear
            self.two_electron_integrals = \
                get_two_electron_integrals_tensor(primitives)

    def init_using_ext(self, primitives):
        size = 6
        gaussian_arr = np.zeros([size*len(primitives)], order='C',
                                dtype=np.double)
        # gaussian_arr_long = gaussian_arr.view(dtype=np.int_)
        gaussian_arr_short = gaussian_arr.view(dtype=np.short)
        for i, g in enumerate(primitives):
            # Write the orbital exponent 
            # and amplitude for the Gaussian3D object
            gaussian_arr[size*i] = g.orbital_exponent()
            gaussian_arr[size*i + 1] = g.amplitude()
            gaussian_arr_short[4*(size*i + 2)] = int(g.angular()[0])
            gaussian_arr_short[4*(size*i + 2) + 1] = int(g.angular()[1])
            gaussian_arr_short[4*(size*i + 2) + 2] = int(g.angular()[2])
            gaussian_arr_short[4*(size*i + 2) + 3] = 0
            gaussian_arr[size*i + 3: size*i + 6] = g.position()
            # Now write the members of each of the individual Gaussian1Ds
            # offset = size*i + 2
            # g1d_size = 3
            # # x
            # gaussian_arr[offset] = g.position()[0]
            # gaussian_arr_long[offset + 1] = int(g.angular()[0])
            # gaussian_arr[offset + 2] = g.orbital_exponent()
            # # y
            # gaussian_arr[offset + g1d_size] =g.position()[1]
            # gaussian_arr_long[offset 
            #                     + g1d_size + 1] = int(g.angular()[1])
            # gaussian_arr[offset + g1d_size + 2] = g.orbital_exponent()
            # # z
            # gaussian_arr[offset + 2*g1d_size] = g.position()[2]
            # gaussian_arr_long[offset
            #                     + 2*g1d_size + 1] = int(g.angular()[2])
            # gaussian_arr[offset + 2*g1d_size + 2] = g.orbital_exponent()
        # print(gaussian_arr)
        # print(gaussian_arr_long)
        nuclear_config = self.nuclear_configuration
        nuc_loc = np.zeros([3*len(nuclear_config)],
                           dtype=np.double, order='C')
        charges = np.zeros([len(nuclear_config)],
                           dtype=np.intc, order='C')
        for i, e in enumerate(self.nuclear_configuration):
            nuc_loc[3*i: 3*i + 3] = e[0]
            charges[i] = e[1]
        print(nuc_loc)
        print(charges)
        self.overlap = np.zeros(2*[len(primitives)],
                                dtype=np.double, order='C')
        self.kinetic = np.zeros(2*[len(primitives)],
                                dtype=np.double, order='C')
        self.nuclear = np.zeros(2*[len(primitives)],
                                dtype=np.double, order='C')
        self.two_electron_integrals = np.zeros(
            4*[len(primitives)], dtype=np.double, order='C')
        extension.compute_overlap_from_primitives(self.overlap, gaussian_arr)
        extension.compute_kinetic_from_primitives(self.kinetic, gaussian_arr)
        extension.compute_nuclear_from_primitives(self.nuclear,
                                                  nuc_loc, charges,
                                                  gaussian_arr)
        extension.compute_two_electron_integrals_from_primitives(
            self.two_electron_integrals, gaussian_arr)
        self.h = self.kinetic + self.nuclear
        # import matplotlib.pyplot as plt
        # plt.imshow(self.overlap)
        # plt.show(); plt.close()
        # plt.imshow(self.kinetic)
        # plt.show(); plt.close()
        # plt.imshow(self.nuclear)
        # plt.show(); plt.close()

    def solve(self, iter_count):
        orbitals = self.orbitals
        energies = self.energies
        two_electron_integrals = self.two_electron_integrals
        h = self.h
        for i in range(iter_count):
            repulsion = 2.0*np.einsum('ijkl,mk,ml->ij',
                                      two_electron_integrals,
                                      orbitals, orbitals,
                                      optimize='greedy',
                                      )
            exchange = np.einsum('ijkl,mj,ml->ik',
                                 two_electron_integrals,
                                 orbitals, orbitals,
                                 optimize='greedy',
                                 )
            fock = h + repulsion - exchange
            energies, eigenvectors = eigh(
                fock, b=self.overlap,
                subset_by_index=[0, self.orbitals_count-1])
            orbitals = eigenvectors.T
            # orbitals =  orbitals/np.sqrt(np.sum(orbitals @ (self.overlap @ orbitals.T)))
        self.orbitals = orbitals
        self.energies = energies
        return energies, orbitals

    def get_kinetic_energy(self):
        return 2.0*np.einsum('ij,ni,nj->', self.kinetic,
                             * 2*[self.orbitals])

    def get_nuclear_potential_energy(self):
        return 2.0*np.einsum('ij,ni,nj->', self.nuclear,
                             * 2*[self.orbitals])

    def get_repulsion_exchange_energy(self):
        repulsion_e = np.einsum('ijkl,ni,nj,mk,ml->',
                                self.two_electron_integrals,
                                * 4*[self.orbitals],
                                optimize='greedy')
        exchange_e = np.einsum('ijkl,ni,mj,nk,ml->',
                               self.two_electron_integrals,
                               * 4*[self.orbitals],
                               optimize='greedy')
        return 2.0*repulsion_e - exchange_e

    def get_nuclear_configuration_energy(self):
        val = 0.0
        if len(self.nuclear_configuration) > 1:
            for i, c_i in enumerate(self.nuclear_configuration):
                for j in range(i+1, len(self.nuclear_configuration)):
                    c_j = self.nuclear_configuration[j]
                    r_i, q_i = c_i
                    r_j, q_j = c_j
                    r = np.linalg.norm(r_i - r_j)
                    val += q_i*q_j/r
            return val
        else:
            return 0.0

    def get_total_energy(self):
        return (self.get_kinetic_energy()
                + self.get_nuclear_potential_energy()
                + self.get_repulsion_exchange_energy()
                + self.get_nuclear_configuration_energy())

File: gaussian_basis/gaussian_basis/src/closed_shell_system.cpp
#include "closed_shell_system.hpp"

#include "Eigen/Core"
#include "Eigen/Sparse"
#include "Eigen/Eigenvalues"

/*
#include <Eigen/Core>
#include <Eigen/Sparse>
#include <Eigen/Eigenvalues>
*/

static inline double at_ind(const double *m, int n, int i, int j) {
    return m[n*i + j];
}


static inline double at_ind(const double *m, int n, int i, int j, int k) {
    return m[n*n*i + n*j + k];
}


static inline double at_ind(
    const double *m, int n, int i, int j, int k, int l_) {
    return m[n*n*n*i + n*n*j + n*k + l_];
}

static inline void reduce_ijkl_mk_ml_to_ij(
    double *dst,
    const double *tensor_ijkl, 
    const double *tensor_mk, const double *tensor_ml,
    int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            for (int m = 0; m < size; m++) {
                for (int k = 0; k < size; k++) {
                    for (int l_ = 0; l_ < size; l_++) {
                        dst[i*size + j] 
                            += at_ind(tensor_ijkl, size, i, j, k, l_)
                            *at_ind(tensor_mk, size, m, k)
                            *at_ind(tensor_ml, size, m, l_);
                    }
                }
            }
        }
    }
}

static inline void reduce_ijkl_mj_ml_to_ij(
    double *dst,
    const double *tensor_ijkl, 
    const double *tensor_mk, const double *tensor_ml,
    int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            for (int m = 0; m < size; m++) {
                for (int k = 0; k < size; k++) {
                    for (int l_ = 0; l_ < size; l_++) {
                        dst[i*size + j] 
                            += at_ind(tensor_ijkl, size, i, j, k, l_)
                            *at_ind(tensor_mk, size, m, j)
                            *at_ind(tensor_ml, size, m, l_);
                    }
                }
            }
        }
    }
}

static inline void add(
    double *dst, 
    const double *a, const double *b, const double *c, int n) {
    for (int i = 0; i < n; i++)
        dst[i] = a[i] + b[i] + c[i];
}

ClosedShellSystem::
ClosedShellSystem(
    int number_of_electrons, 
    const Orbitals &orbitals,
    const NuclearConfiguration &nuclear_config) {
    this->nuclear_configuration = nuclear_config;
    this->kinetic = orbitals.get_kinetic();
    this->nuclear = orbitals.get_nuclear_potential(nuclear_configuration);
    for (int i = 0; i < this->nuclear.size(); i++)
        this->hamiltonian[i] = this->kinetic[i] + this->nuclear[i];
    this->two_electron_integrals = orbitals.get_two_electron_integrals();
    this->energies = std::vector<double>(
        orbitals.number_of_basis_functions(), 0.0);
    // for (const AtomicOrbitalsParameters &atomic_orbitals_params: orbitals) {

    // }
}

typedef Eigen::MatrixXd Matrix;

void ClosedShellSystem::solve(int iter_count) {
    const double *orbitals = &this->orbitals[0];
    const double *energies = &this->energies[0];
    const double *two_electron_integrals 
        = &this->two_electron_integrals[0];
    const double *hamiltonian = &this->hamiltonian[0];
    int n = this->basis_func_count;
    for (int i = 0; i < iter_count; i++) {
        // std::vector<double> fock(n*n, 0.0);
        Matrix fock = Matrix::Zero(1, n*n);
        std::vector<double> repulsion(n*n, 0.0);
        std::vector<double> exchange(n*n, 0.0);
        reduce_ijkl_mk_ml_to_ij(
            &repulsion[0], 
            two_electron_integrals, 
            orbitals, orbitals, n);
        reduce_ijkl_mj_ml_to_ij(
            &exchange[0], 
            two_electron_integrals, 
            orbitals, orbitals, n);
        add(&fock(0, 0), hamiltonian, &repulsion[0], &exchange[0], n*n);
        fock.resize(n, n);
        Eigen::EigenSolver<Matrix> solver(fock);
        Matrix eigenvectors = solver.eigenvectors().transpose();
        eigenvectors.resize(1, n*n);
        Matrix eigenvalues = solver.eigenvalues();
        for (int i = 0; i < n; i++) {
            this->energies[i] = eigenvalues[i];
            for (int j = 0; j < n; j++)
                this->orbitals[i*n + j] = eigenvectors(i, j);
        }

    }
}

double ClosedShellSystem::get_kinetic_energy() const {

}

double ClosedShellSystem::get_nuclear_potential_energy() const {

}

double ClosedShellSystem::get_repulsion_exchange_energy() const {

}

double ClosedShellSystem::get_nuclear_configuration_energy() const {
    
}

double ClosedShellSystem::get_total_energy() const {
    return this->get_kinetic_energy() + 
        this->get_nuclear_potential_energy() +
        this->get_repulsion_exchange_energy() +
        this->get_nuclear_configuration_energy();
}
